pkg/server/api/svid/v1/service_test.go

```diff
@@ -921,10 +926,206 @@ func TestServiceBatchNewX509SVID(t *testing.T) {
+			entry:          downstreamEntryUnauthorized,
+		},
+		{
+			name:           "Nonexistent entry",
+			rateLimiterErr: nil,
+			err:            "error not found or not authorized",
+			failSigning:    false,
+			csr:            []byte{1, 2, 3},
+			code:           codes.NotFound,
+			fetcherErr:     "",
+			entry:          &types.Entry{Id: "non-existent-entry"},
+		},
+		{
+			name:           "Entry fetch error",
+			rateLimiterErr: nil,
+			err:            "failed to fetch registration entries",
+			failSigning:    false,
+			csr:            []byte{1, 2, 3},
+			code:           codes.Internal,
+			fetcherErr:     "Entry fetch error",
+			entry:          downstreamEntry,
+		},
+		{
+			name:           "FailDataStore",
+			rateLimiterErr: nil,
+			err:            "bundle not found",
+			failSigning:    false,
+			csrTemplate:    &x509.CertificateRequest{},
+			code:           codes.Internal,
+			fetcherErr:     "",
+			entry:          downstreamEntry,
+			failDataStore:  true,
+		},
+		{
+			name:           "CertRequestWrongTd",
+			rateLimiterErr: nil,
+			err:            "error not found or not authorized",
+			failSigning:    false,
+			failDataStore:  false,
+			csrTemplate: &x509.CertificateRequest{
+				URIs: []*url.URL{workloadID.URL()},
+			},
+			fetcherErr: "",
+			entry:      downstreamEntryWrongTd,
+			code:       codes.NotFound,
+		},
+		{
+			name:           "SuccessfulCARequest",
+			rateLimiterErr: nil,
+			err:            "",
+			failSigning:    false,
+			failDataStore:  false,
+			csrTemplate: &x509.CertificateRequest{
+				// Right now, the request will succeed even if URIs is empty. This is
+				// probably not the right behavior.
+				URIs: []*url.URL{workloadID.URL()},
+			},
+			fetcherErr: "",
+			entry:      downstreamEntry,
+		},
+	} {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			test.logHook.Reset()
+			test.ef.err = tt.fetcherErr
+			if !tt.failDataStore {
+				_, err := test.ds.AppendBundle(context.TODO(), &datastore.AppendBundleRequest{
+					Bundle: &common.Bundle{
+						// The SPIFFE ID of the bundle in the datastore needs to match the SPIFFE ID
+						// provided by the client
+						TrustDomainId: td.IDString(),
+						RootCas: []*common.Certificate{
+							{DerBytes: []byte("RootCa1")},
+						},
+					},
+				})
+				require.NoError(t, err)
+			} else {
+				_, err := test.ds.DeleteBundle(context.TODO(),
+					&datastore.DeleteBundleRequest{TrustDomainId: td.IDString()},
+				)
+				require.NoError(t, err)
+			}
+
+			test.withCallerID = true
+			test.rateLimiter.count = 1
+			test.rateLimiter.err = tt.rateLimiterErr
+			test.ca.SetX509CA(x509CA)
+			if tt.failSigning {
+				test.ca.SetX509CA(nil)
+			}
+
+			csr := tt.csr
+			if tt.csrTemplate != nil {
+				csr = createCSR(t, tt.csrTemplate)
+			}
+
+			ctx := context.Background()
+			resp, err := test.client.NewDownstreamX509CA(ctx, &svidpb.NewDownstreamX509CARequest{
+				EntryId: tt.entry.Id,
+				Csr:     csr,
+			})
+
+			if tt.err == "" {
```

Comment:
nit: maybe invert the conditions and return early so we don't have so much nested code?
```suggestion
			if tt.err != "" {
			    spiretest.RequireGRPCStatusContains(....)
			    return
			}
			
			.... bunch of assertions around response contents...
```